### PR TITLE
batch state get

### PR DESCRIPTION
**Pull request by pared**
* [ ] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [ ] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----
Related to #2093

### CODE REVIEW

**Comment by pared** [comment]
Some benchmark for current change:

Test case:
- add 100K files in signle directory
- status (just to trigger unpacked dir creation, so it won't take time later)
- remove one file (to enforce `_collect_dir`)
- status - this one execution time is measured

On master, where we get one by one:
34 seconds for status, in which 17 seconds we spend in `state.get`

This change:
30 seconds, in which 10 seconds we spend in `state.get_multiple`

**Comment by efiop** [comment]
@pared Just a heads up: a lot of tests failed

**Comment by Suor** [comment]
Several comments on this:
1. The gain is small, not worth the effort and code complication for me.
2. We can try `cursor.executemany()` or using `WITH clauses` to batch inserts and updates too.
3. Switching to some key/value db like LevelDB or KyotoCabinet might be a way forward.

**Comment by Suor** [comment]
Also, having `.get()` and `.get_single()` looks weird. `.get()` sounds single by itself.

File: dvc/state.py [hunk 1]
```diff
@@ -37,7 +37,10 @@ class StateNoop(object):
     def save(self, path_info, checksum):
         pass
 
-    def get(self, path_info):
+    def get(self, path_infos):
+        return {pi: None for pi in path_infos}
+
+    def get_single(self, path_info):
```
Comment (efiop) [inline dvc/state.py]: Agree with @Suor . Let's remove this, and use `get([info])` when we need it.

**Comment by pared** [comment]
@efiop What do you think about @Suor comment relating to complexity versus potential gain? I have to admit that this change makes get method quite complicated.

**Comment by efiop** [comment]
@pared @Suor 

> The gain is small, not worth the effort and code complication for me.

shaving 4 sec from 34 sec operation is significant and is worth this change, especially since we already have it here.

> We can try cursor.executemany() or using WITH clauses to batch inserts and updates too.

inserts and updates are out of scope for this PR. @pared did you consider `executemany`?

> Switching to some key/value db like LevelDB or KyotoCabinet might be a way forward.

@Suor Feel free to create an issue and share your experience with those.

**Comment by pared** [comment]
>have you considered executemany

@efiop I did not, but it doesn't seem like big change, ill try to modify code and compare performance

File: dvc/state.py (hunk 1 above)
Comment (Suor) [reply dvc/state.py]: Or make `.get()` accept both lists and single path infos.

Comment (Suor) [reply dvc/state.py]: Or make .get() accept both lists and single path infos.

**Comment by Suor** [comment]
> shaving 4 sec from 34 sec operation is significant and is worth this change, especially since we already have it here.

And we will live with it. If we really need to speed this up we should look elsewhere, 11% speed gain won't make or break it for anyone.

**Comment by efiop** [comment]
> And we will live with it, will need to support it. If we really need to speed this up we should look elsewhere, 11% speed gain won't make or break it for anyone.

I don't see anything that scary about this patch. 12% gain is not a breakthrough, but it is nice to have.

**Comment by shcheklein** [comment]
Btw, when we do these tests do we populate the DB to 10M limit at least? It might change 12% gain to one or another.

**Comment by pared** [comment]
@shcheklein The tests were performed for 100k elements, so number of entries was approximately the same.

**Comment by pared** [comment]
@efiop @Suor I redesigned `get_multiple` a little bit. Could you take a look now?

**Comment by shcheklein** [comment]
@pared yes, my point that there is a chance that we can save more time if DB is populated at its full capacity.

can you also show the full dump of the profiler?

I think these changes are so well incapsulated that it should not hurt us much long time (in terms of supporting this).

**Comment by pared** [comment]
@shcheklein Ill adjust test accordingly

**Comment by pared** [comment]
Note: I should also batch timestamp inserts

**Comment by pared** [comment]
@shcheklein I performed test with 100k files, and 10M random entries inside state database, I include pstats for first 50 calls, sorted by cumtime.
master
```
Wed Aug 14 11:16:52 2019    master

         54461731 function calls (54030551 primitive calls) in 33.419 seconds

   Ordered by: cumulative time
   List reduced from 8126 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2488/1    0.024    0.000   33.422   33.422 {built-in method builtins.exec}
        1    0.000    0.000   33.422   33.422 time_status.py:1(<module>)
        1    0.000    0.000   31.490   31.490 status.py:94(status)
        1    0.000    0.000   31.467   31.467 status.py:10(_local_status)
        1    0.000    0.000   30.904   30.904 stage.py:867(status)
        2    0.000    0.000   30.903   15.452 stage.py:858(_status)
        1    0.000    0.000   30.903   30.903 base.py:180(status)
      3/2    0.033    0.011   29.480   14.740 base.py:271(get_checksum)
        1    0.000    0.000   29.479   29.479 base.py:166(changed_checksum)
        1    0.000    0.000   29.479   29.479 base.py:300(save_info)
        1    0.077    0.077   26.896   26.896 base.py:202(get_dir_checksum)
        1    0.001    0.001   23.623   23.623 base.py:164(_collect_dir)
   100002    0.478    0.000   16.878    0.000 state.py:386(get)
        1    0.234    0.234   14.325   14.325 base.py:174(<dictcomp>)
   200010    0.166    0.000    9.456    0.000 path_info.py:39(__str__)
   200024    0.379    0.000    9.000    0.000 __init__.py:405(relpath)
        1    0.232    0.232    6.737    6.737 base.py:184(<listcomp>)
   100005    0.662    0.000    6.528    0.000 fs.py:23(get_mtime_and_size)
   100005    0.188    0.000    6.219    0.000 fs.py:17(get_inode)
   600137    0.799    0.000    6.038    0.000 posixpath.py:376(abspath)
   200024    0.920    0.000    5.834    0.000 posixpath.py:453(relpath)
   402837    0.576    0.000    5.452    0.000 {method 'format' of 'str' objects}
   100001    0.124    0.000    4.727    0.000 pathlib.py:704(as_posix)
       89    0.002    0.000    4.232    0.048 __init__.py:1(<module>)
   710971    2.505    0.000    3.082    0.000 {built-in method posix.stat}
   600145    2.035    0.000    3.052    0.000 posixpath.py:338(normpath)
   100005    0.234    0.000    2.839    0.000 state.py:344(get_state_record_for_inode)
   200017    0.188    0.000    2.732    0.000 state.py:133(_execute)
   300016    0.382    0.000    2.650    0.000 pathlib.py:631(_parse_args)
        7    0.060    0.009    2.371    0.339 {method 'update' of 'set' objects}
   200017    2.327    0.000    2.327    0.000 {method 'execute' of 'sqlite3.Cursor' objects}
        3    0.007    0.002    2.236    0.745 state.py:357(save)
   300017    1.272    0.000    2.153    0.000 pathlib.py:51(parse_parts)
   902427    1.423    0.000    2.140    0.000 posixpath.py:75(join)
        1    0.001    0.001    2.076    2.076 base.py:215(_get_dir_info_checksum)
   1610/9    0.009    0.000    2.074    0.230 <frozen importlib._bootstrap>:966(_find_and_load)
   1610/9    0.007    0.000    2.074    0.230 <frozen importlib._bootstrap>:936(_find_and_load_unlocked)
  1528/10    0.008    0.000    2.073    0.207 <frozen importlib._bootstrap>:651(_load_unlocked)
  1438/10    0.004    0.000    2.072    0.207 <frozen importlib._bootstrap_external>:672(exec_module)
  1874/10    0.001    0.000    2.070    0.207 <frozen importlib._bootstrap>:211(_call_with_frames_removed)
        1    0.253    0.253    2.050    2.050 __init__.py:120(dump)
   100000    0.106    0.000    1.998    0.000 base.py:172(<genexpr>)
   100001    0.174    0.000    1.948    0.000 state.py:295(_update_state_record_timestamp_for_inode)
   200009    0.067    0.000    1.892    0.000 pathlib.py:898(__truediv__)
        5    0.001    0.000    1.858    0.372 __init__.py:113(dict_md5)
   654/28    0.002    0.000    1.828    0.065 {built-in method builtins.__import__}
   200009    0.208    0.000    1.825    0.000 pathlib.py:685(_make_child)
    99999    0.320    0.000    1.751    0.000 pathlib.py:846(relative_to)
   400001    0.185    0.000    1.459    0.000 __init__.py:319(walk_files)
        1    0.000    0.000    1.423    1.423 base.py:174(changed_cache)
```

2093 (this change)
```
Wed Aug 14 11:27:25 2019    2093

         53163448 function calls (52732266 primitive calls) in 31.760 seconds

   Ordered by: cumulative time
   List reduced from 8146 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2488/1    0.021    0.000   31.764   31.764 {built-in method builtins.exec}
        1    0.000    0.000   31.763   31.763 time_status.py:1(<module>)
        1    0.000    0.000   30.672   30.672 status.py:94(status)
        1    0.000    0.000   30.625   30.625 status.py:10(_local_status)
        1    0.000    0.000   30.008   30.008 stage.py:867(status)
        2    0.000    0.000   30.007   15.004 stage.py:858(_status)
        1    0.000    0.000   30.007   30.007 base.py:180(status)
      3/2    0.030    0.010   28.689   14.344 base.py:271(get_checksum)
        1    0.000    0.000   28.688   28.688 base.py:166(changed_checksum)
        1    0.000    0.000   28.688   28.688 base.py:300(save_info)
        1    0.109    0.109   26.212   26.212 base.py:202(get_dir_checksum)
        1    0.011    0.011   22.660   22.660 base.py:164(_collect_dir)
        1    0.443    0.443   11.342   11.342 state.py:411(get_multiple)
   200010    0.157    0.000    8.976    0.000 path_info.py:39(__str__)
   200024    0.339    0.000    8.508    0.000 __init__.py:405(relpath)
        1    0.244    0.244    7.100    7.100 base.py:184(<listcomp>)
   100005    0.646    0.000    6.486    0.000 fs.py:23(get_mtime_and_size)
   600137    0.762    0.000    5.773    0.000 posixpath.py:376(abspath)
   200024    0.838    0.000    5.583    0.000 posixpath.py:453(relpath)
        1    0.067    0.067    5.531    5.531 state.py:423(<dictcomp>)
   100005    0.155    0.000    5.465    0.000 fs.py:17(get_inode)
   100001    0.131    0.000    4.921    0.000 pathlib.py:704(as_posix)
   203045    0.304    0.000    4.515    0.000 {method 'format' of 'str' objects}
        7    0.108    0.015    3.944    0.563 {method 'update' of 'set' objects}
   300016    0.573    0.000    3.730    0.000 pathlib.py:631(_parse_args)
   100000    0.197    0.000    3.328    0.000 base.py:172(<genexpr>)
   200009    0.103    0.000    3.131    0.000 pathlib.py:898(__truediv__)
   200009    0.334    0.000    3.028    0.000 pathlib.py:685(_make_child)
   300017    1.713    0.000    2.974    0.000 pathlib.py:51(parse_parts)
   600145    1.974    0.000    2.974    0.000 posixpath.py:338(normpath)
        3    0.008    0.003    2.626    0.875 state.py:438(get)
   710975    2.114    0.000    2.523    0.000 {built-in method posix.stat}
       89    0.002    0.000    2.500    0.028 __init__.py:1(<module>)
        3    0.005    0.002    2.307    0.769 state.py:357(save)
        1    0.001    0.001    2.274    2.274 base.py:215(_get_dir_info_checksum)
        1    0.278    0.278    2.243    2.243 __init__.py:120(dump)
   100001    0.169    0.000    2.077    0.000 state.py:295(_update_state_record_timestamp_for_inode)
        5    0.001    0.000    2.045    0.409 __init__.py:113(dict_md5)
   902427    1.284    0.000    1.998    0.000 posixpath.py:75(join)
    99999    0.331    0.000    1.906    0.000 pathlib.py:846(relative_to)
   400001    0.176    0.000    1.510    0.000 __init__.py:319(walk_files)
   100119    0.107    0.000    1.376    0.000 state.py:133(_execute)
        1    0.000    0.000    1.319    1.319 base.py:174(changed_cache)
        1    0.000    0.000    1.319    1.319 base.py:617(changed_cache)
        1    0.000    0.000    1.319    1.319 base.py:602(_changed_dir_cache)
        1    0.000    0.000    1.318    1.318 __init__.py:561(_changed_unpacked_dir)
        5    0.001    0.000    1.309    0.262 __init__.py:183(dumps)
        5    0.000    0.000    1.309    0.262 encoder.py:182(encode)
        6    1.288    0.215    1.288    0.215 encoder.py:204(iterencode)
   1610/9    0.007    0.000    1.232    0.137 <frozen importlib._bootstrap>:966(_find_and_load)
```

The difference in execution time is not as visible as it was in previous test. That might be caused by new implementation.

**Comment by pared** [comment]
Also it seems to me that I made a mistake by interpreting performance improvement as a result of batching the calls. As you can see in the cProfile logs, cumulative time spend in `state._execute` differs by 1 second, that would suggest that we don't save as much time on sql calls, but rather implementation makes a difference.

**Comment by efiop** [comment]
Also it is interesting how long abspath, relpath, `__str__` take. Like the majority of the time.

**Comment by pared** [comment]
@efiop I guess thats a result of number of files that need to be checked, maybe we should actually think how to minimaize number of calls to this simple methods.

**Comment by efiop** [comment]
@pared So is the new PR version slower than the old one? Was the old one broken? If the difference is in 1 second, then this PR is not worth the effort :slightly_frowning_face:

**Comment by pared** [comment]
@efiop new version is (at least for me) more readable, I can roll back to previous one and perform more in depth analysis of that one, maybe we will get some more insight out of that. How about that?

**Comment by efiop** [comment]
@pared what is the baseline though? You did some tests with regular gets and batching without dvc a while ago, were those tests broken too?

**Comment by pared** [comment]
@efiop good point. I found those test, and they are not broken. The difference was number of selects. This test assumed collecting information about dir of 100k files. My "pure sql" tests were obtaining 1M files. 

I have rerun the sql test for 100k and difference is 2.5 s for getting one by one versus 0.05 s for getting in bulk. The same test for 1M files yielded respectively 25s versus 0.5s. 

I will run 1M files directory test for dvc.

**Comment by pared** [comment]
Ok, i did some tests for 1M, and here are the results:
Test environment:
- `dvc add` for `data` dir with 1M small files in it
- `state` db with artificial 10M entries
- remove one file from `data`
- check status (removing one file will trigger collecting whole dir and comparing it with state db entries)

results for master (as of friday, 16.08.2019):
```
Fri Aug 16 10:59:16 2019    master_1M

         534166016 function calls (530134836 primitive calls) in 309.912 seconds

   Ordered by: cumulative time
   List reduced from 8126 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2488/1    0.020    0.000  309.915  309.915 {built-in method builtins.exec}
        1    0.000    0.000  309.915  309.915 time_status.py:1(<module>)
        1    0.000    0.000  308.842  308.842 status.py:94(status)
        1    0.000    0.000  308.798  308.798 status.py:10(_local_status)
        1    0.000    0.000  304.723  304.723 stage.py:867(status)
        2    0.000    0.000  304.722  152.361 stage.py:858(_status)
        1    0.000    0.000  304.722  304.722 base.py:180(status)
      3/2    0.416    0.139  291.482  145.741 base.py:271(get_checksum)
        1    0.000    0.000  291.482  291.482 base.py:166(changed_checksum)
        1    0.000    0.000  291.482  291.482 base.py:300(save_info)
        1    0.846    0.846  266.389  266.389 base.py:202(get_dir_checksum)
        1    0.010    0.010  232.604  232.604 base.py:164(_collect_dir)
  1000002    4.557    0.000  163.593    0.000 state.py:386(get)
        1    2.289    2.289  141.096  141.096 base.py:174(<dictcomp>)
  2000010    1.733    0.000   91.591    0.000 path_info.py:39(__str__)
  2000024    3.645    0.000   86.921    0.000 __init__.py:405(relpath)
        1    2.405    2.405   66.065   66.065 base.py:184(<listcomp>)
  1000005    6.553    0.000   64.505    0.000 fs.py:23(get_mtime_and_size)
  1000005    1.786    0.000   60.523    0.000 fs.py:17(get_inode)
  6000137    7.654    0.000   58.069    0.000 posixpath.py:376(abspath)
  2000024    8.833    0.000   56.168    0.000 posixpath.py:453(relpath)
  4002866    5.828    0.000   52.873    0.000 {method 'format' of 'str' objects}
  1000001    1.241    0.000   46.021    0.000 pathlib.py:704(as_posix)
  6000145   19.948    0.000   29.722    0.000 posixpath.py:338(normpath)
  1000005    2.318    0.000   28.748    0.000 state.py:344(get_state_record_for_inode)
  2000017    1.979    0.000   27.544    0.000 state.py:133(_execute)
  7011051   20.441    0.000   26.253    0.000 {built-in method posix.stat}
  3000016    3.849    0.000   25.713    0.000 pathlib.py:631(_parse_args)
        3    0.057    0.019   24.883    8.294 state.py:357(save)
  2000017   23.444    0.000   23.444    0.000 {method 'execute' of 'sqlite3.Cursor' objects}
        7    0.638    0.091   23.207    3.315 {method 'update' of 'set' objects}
        5    0.016    0.003   22.335    4.467 __init__.py:113(dict_md5)
  3000017   12.267    0.000   20.759    0.000 pathlib.py:51(parse_parts)
        1    0.000    0.000   20.487   20.487 base.py:215(_get_dir_info_checksum)
  9002446   13.435    0.000   20.446    0.000 posixpath.py:75(join)
        1    2.496    2.496   20.308   20.308 __init__.py:120(dump)
  1000001    1.770    0.000   19.195    0.000 state.py:295(_update_state_record_timestamp_for_inode)
  1000000    1.012    0.000   18.904    0.000 base.py:172(<genexpr>)
  2000009    0.657    0.000   17.892    0.000 pathlib.py:898(__truediv__)
   999999    3.135    0.000   17.356    0.000 pathlib.py:846(relative_to)
  2000009    1.969    0.000   17.235    0.000 pathlib.py:685(_make_child)
        5    0.012    0.002   15.474    3.095 __init__.py:183(dumps)
        5    0.000    0.000   15.462    3.092 encoder.py:182(encode)
        6   15.267    2.544   15.267    2.544 encoder.py:204(iterencode)
  4000001    1.645    0.000   14.406    0.000 __init__.py:319(walk_files)
        1    0.000    0.000   13.240   13.240 base.py:174(changed_cache)
        1    0.000    0.000   13.240   13.240 base.py:616(changed_cache)
        1    0.000    0.000   13.240   13.240 base.py:601(_changed_dir_cache)
        1    0.000    0.000   13.240   13.240 __init__.py:565(_changed_unpacked_dir)
  9999992    1.538    0.000   10.637    0.000 encoder.py:412(_iterencode)
```

results for my change:
```
Fri Aug 16 11:44:17 2019    2093_1M

         531180162 function calls (527148982 primitive calls) in 289.144 seconds

   Ordered by: cumulative time
   List reduced from 8136 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2488/1    0.021    0.000  289.147  289.147 {built-in method builtins.exec}
        1    0.000    0.000  289.147  289.147 time_status.py:1(<module>)
        1    0.000    0.000  288.069  288.069 status.py:94(status)
        1    0.000    0.000  288.031  288.031 status.py:10(_local_status)
        1    0.000    0.000  283.812  283.812 stage.py:867(status)
        2    0.000    0.000  283.812  141.906 stage.py:858(_status)
        1    0.000    0.000  283.812  283.812 base.py:180(status)
      3/2    0.419    0.140  268.130  134.065 base.py:271(get_checksum)
        1    0.000    0.000  268.129  268.129 base.py:166(changed_checksum)
        1    0.000    0.000  268.129  268.129 base.py:300(save_info)
        1    0.880    0.880  242.921  242.921 base.py:202(get_dir_checksum)
        1    0.395    0.395  209.602  209.602 base.py:164(_collect_dir)
        4    4.495    1.124  146.753   36.688 state.py:421(get_multiple)
  2000010    1.592    0.000   83.524    0.000 path_info.py:39(__str__)
  2000024    3.168    0.000   79.017    0.000 __init__.py:405(relpath)
        1    2.174    2.174   65.191   65.191 base.py:184(<listcomp>)
  1000005    5.596    0.000   64.381    0.000 fs.py:23(get_mtime_and_size)
  6000137    6.979    0.000   53.441    0.000 posixpath.py:376(abspath)
        4    1.458    0.365   53.024   13.256 state.py:451(<dictcomp>)
  2000024    7.823    0.000   51.932    0.000 posixpath.py:453(relpath)
  1000005    1.485    0.000   51.183    0.000 fs.py:17(get_inode)
  1000001    1.367    0.000   45.838    0.000 pathlib.py:704(as_posix)
  2004870    3.079    0.000   42.362    0.000 {method 'format' of 'str' objects}
        4    1.487    0.372   41.749   10.437 state.py:446(<dictcomp>)
  1000001    1.946    0.000   28.805    0.000 state.py:298(_update_state_record_timestamp_for_inode)
        3    0.000    0.000   28.280    9.427 state.py:471(get)
  6000145   18.263    0.000   27.470    0.000 posixpath.py:338(normpath)
  3000016    3.833    0.000   25.536    0.000 pathlib.py:631(_parse_args)
        3    0.054    0.018   24.471    8.157 state.py:392(save)
  7011051   20.751    0.000   23.762    0.000 {built-in method posix.stat}
        7    0.649    0.093   23.323    3.332 {method 'update' of 'set' objects}
        5    0.016    0.003   22.215    4.443 __init__.py:113(dict_md5)
  3000017   12.066    0.000   20.574    0.000 pathlib.py:51(parse_parts)
        1    0.001    0.001   20.159   20.159 base.py:215(_get_dir_info_checksum)
        1    2.488    2.488   19.974   19.974 __init__.py:120(dump)
  1001019    1.072    0.000   19.848    0.000 state.py:136(_execute)
  9002446   12.227    0.000   19.105    0.000 posixpath.py:75(join)
  1000000    1.023    0.000   18.995    0.000 base.py:172(<genexpr>)
  2000009    0.652    0.000   17.972    0.000 pathlib.py:898(__truediv__)
  1001019   17.347    0.000   17.347    0.000 {method 'execute' of 'sqlite3.Cursor' objects}
  2000009    1.939    0.000   17.320    0.000 pathlib.py:685(_make_child)
   999999    3.042    0.000   16.906    0.000 pathlib.py:846(relative_to)
        1    0.000    0.000   15.682   15.682 base.py:174(changed_cache)
        1    0.000    0.000   15.682   15.682 base.py:616(changed_cache)
        1    0.000    0.000   15.682   15.682 base.py:601(_changed_dir_cache)
        1    0.000    0.000   15.681   15.681 __init__.py:565(_changed_unpacked_dir)
        5    0.012    0.002   15.384    3.077 __init__.py:183(dumps)
        5    0.000    0.000   15.372    3.074 encoder.py:182(encode)
        6   15.173    2.529   15.173    2.529 encoder.py:204(iterencode)
  4000001    1.718    0.000   14.523    0.000 __init__.py:319(walk_files)

```

**Comment by pared** [comment]
Notes:
- against expectation, batching calls did not make that much of a difference  (we saved around 20s out of 300s of whole operation)
- time saved on sql calls is negligible comparing to time of operation (23 vs 17 seconds), that is also contrary to my tests of batching calls for just sql calls (which differed by order of magnitude). I am unable to reason why this happend, would need to invesitgate more.
- previous two comments (and results of tests on 100k objects) suggest that previously found savings of time were more the result of code rafactoring, rather than great improvement thanks to batching the calls . That also indicates that I was wrong and misinterpreted tests results.

Conclusion:
- batching the calls by itself will not be as much of an improvement as we hoped. There is some space to work on optimization, but maybe we should start with simpler tasks (like spending most of time in `abspath`, `relpath`) rather than reorganizing whole state access.
- time savings for previously made changes (batching `state.remove_unsued_links` in https://github.com/iterative/dvc/pull/2194) should be reevaluated, and possibly, this change should be rolled back

**Comment by efiop** [comment]
@pared Ok, so let's close this PR and the issue behind it and revert #2194 (unless there is something useful in it and then we should remove only the unneeded pieces) ?

**Comment by pared** [comment]
Ill look through that PR

**Comment by efiop** [comment]
@pared You could create a separate issue to not forget about it too.

**Pull request closed without merging**